packages/eas-cli/src/branch/queries.ts

```diff
@@ -0,0 +1,114 @@
+import chalk from 'chalk';
+import CliTable from 'cli-table3';
+
+import { PaginatedQueryFlags } from '../commandUtils/flagHelpers';
+import { UpdateBranchFragment } from '../graphql/generated';
+import { BranchQuery } from '../graphql/queries/BranchQuery';
+import Log from '../log';
+import { UPDATE_COLUMNS, formatUpdate, getPlatformsForGroup } from '../update/utils';
+import { printJsonOnlyOutput } from '../utils/json';
+import {
+  PaginatedQueryPromptOptions,
+  PaginatedQueryPromptType,
+  PaginatedQueryResponse,
+  performPaginatedQueryAsync,
+} from '../utils/queries';
+
+export const BRANCHES_LIMIT = 50;
+
+export async function selectBranchForProjectAsync(
+  projectId: string,
+  promptTitle: string,
+  flags: PaginatedQueryFlags
+): Promise<UpdateBranchFragment> {
+  if (flags['non-interactive']) {
+    throw new Error('Unable to select a branch in non-interactive mode.');
+  }
+
+  const fromatUpdateBranchFragment = (updateBranchFragment: UpdateBranchFragment): string =>
+    updateBranchFragment.name;
+  const getIdentifierForQueryItem = (updateBranchFragment: UpdateBranchFragment): string =>
+    updateBranchFragment.id;
+
+  const branch = await performPaginatedQueryAsync({
+    pageSize: flags.limit ?? BRANCHES_LIMIT,
+    offset: flags.offset ?? 0,
+    queryToPerform: queryBranchesForProject(projectId),
+    promptOptions: {
+      type: PaginatedQueryPromptType.select,
+      title: promptTitle,
+      getIdentifierForQueryItem,
+      createDisplayTextForSelectionPromptListItem: fromatUpdateBranchFragment,
+    },
+  });
+  return branch.pop()!;
```

Comment:
> It's bad practice to use ! in TS. For now, it's probably working fine but what if performPaginatedQueryAsync started returning empty list?

This is a good point and something that @wkozyra95 mentioned as well. I think it would make sense to throw an error if the list was empty or update the return type to a `Maybe<T>`. I'll add some handling in the next commit.

> .pop mutates the original list.

Yup. Since it will be garbage collected when the function exits, this is not an issue.

> Just by reading code, it's not clear to me what's the type of branch. Seeing .pop makes me think it's a list but the name is singular.

The return type is specified in the function declaration: `UpdateBranchFragment[]`. I will update the name to be branches! The select logic will return _all_ items that match the `getIdentifierForQueryItem` check, and popping will return the last item in the list (in this case the only possible item in the list).

Reply:
> Yup. Since it will be garbage collected when the function exits, this is not an issue.

It's still an issue. What if someone refactored the function and didn't realize you mutated the list? There is even a concept of immutable objects because this is such a common problem in JS world, see https://immutable-js.com/

Reply:
yes yes! haha, immutability is a thing (another tenant of functional programming). 

> What if someone refactored the function and didn't realize you mutated the list?

I would hope whoever contributes understands the code they are re-writing 😛 It seems a pretty unlikely scenario to me but I did changes this as per your request as it's nbd.